storage/memetcd/peer_store.go

```diff
@@ -0,0 +1,574 @@
+	etcdOp := func() error {
+		err := ps.etcdDeletePeer(ih, p, true)
+		if err == storage.ErrResourceDoesNotExist {
+			log.Error(fmt.Sprintf("deleteseeder: announce for infohash=%v, peer=%v does not exist in etcd", ih, p.ID))
+			return nil
+		}
+		return err
+	}
+	return ps.doOp(memOp, etcdOp)
+}
+
+func (ps *peerStore) PutLeecher(ih bittorrent.InfoHash, p bittorrent.Peer) error {
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped memory store")
+	default:
+	}
+
+	memOp := func() error {
+		return ps.memoryStore.PutLeecher(ih, p)
+	}
+	etcdOp := func() error {
+		return ps.etcdPutOrUpdatePeer(ih, p, false)
+	}
+	return ps.doOp(memOp, etcdOp)
+}
+
+func (ps *peerStore) DeleteLeecher(ih bittorrent.InfoHash, p bittorrent.Peer) error {
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped memory store")
+	default:
+	}
+
+	memOp := func() error {
+		return ps.memoryStore.DeleteLeecher(ih, p)
+	}
+	etcdOp := func() error {
+		err := ps.etcdDeletePeer(ih, p, false)
+		if err == storage.ErrResourceDoesNotExist {
+			log.Error(fmt.Sprintf("deleteleecher: announce for infohash=%v, peer=%v does not exist in etcd", ih, p.ID))
+			return nil
+		}
+		return err
+	}
+	return ps.doOp(memOp, etcdOp)
+}
+
+func (ps *peerStore) GraduateLeecher(ih bittorrent.InfoHash, p bittorrent.Peer) error {
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped memory store")
+	default:
+	}
+
+	memOp := func() error {
+		return ps.memoryStore.GraduateLeecher(ih, p)
+	}
+	etcdOp := func() error {
+		return ps.etcdGraduateLeecher(ih, p)
+	}
+	return ps.doOp(memOp, etcdOp)
+}
+
+func (ps *peerStore) AnnouncePeers(ih bittorrent.InfoHash, seeder bool, numWant int, announcer bittorrent.Peer) (peers []bittorrent.Peer, err error) {
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped memory store")
+	default:
+	}
+
+	peers, err = ps.memoryStore.AnnouncePeers(ih, seeder, numWant, announcer)
+	return
+}
+
+func (ps *peerStore) ScrapeSwarm(ih bittorrent.InfoHash, addressFamily bittorrent.AddressFamily) (resp bittorrent.Scrape) {
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped memory store")
+	default:
+	}
+
+	resp = ps.memoryStore.ScrapeSwarm(ih, addressFamily)
+	return
+}
+
+func (ps *peerStore) Stop() <-chan error {
+	c := make(chan error, 2)
+	go func() {
+		close(ps.closed)
+		if err := <-ps.memoryStore.Stop(); err != nil {
+			c <- err
+		}
+
+		if err := ps.etcd.Close(); err != nil {
+			c <- err
+		}
+		close(c)
+	}()
+
+	return c
+}
+
+func (ps *peerStore) LogFields() log.Fields {
+	return ps.cfg.LogFields()
+}
+
+// RACES IN ETCD OPERATIONS FOR (INFOHASH, PEERKEY) WHEN ETCD ASYNC OP IS TRUE
+// ---------------------------------------------------------------------------
+// There can be a race among deleteseeder, graduateleecher or any other etcdop for same ih,p when ps.cfg.EtcdOpAsync == true
+// This should be a rare event unless torrent clients are misbehaving
```

Comment:
I guess deleteOp can also happen if torrent stopped or deleted. So, in **normal** torrent client behaviour, it might occur that during a regular announce request, if someone stopped the torrent at the same time, two requests(announce req & stopped req) might be interleaved(a possibility of race), so it isn't necessary that clients are misbehaving if the race occurs, no?

Reply:
Agreed, I thought of that later. But, I think we can live with the race for now and would like to validate that in production/tests before fixing. Maybe, as a separate PR?